src/nn/webgl/WebGLModel.js

```diff
@@ -0,0 +1,328 @@
+        }
+      } break;
+      case OperationCode.AVERAGE_POOL_2D:
+      case OperationCode.MAX_POOL_2D: {
+        let inCount = inputs.length;
+        if (inCount !== 7 && inCount !== 10) {
+          throw new Error(`Invalid parameters number of Pooling ${op}`);
+        }
+        let i = 0;
+        let input = operands[inputs[i++]];
+        let output = operands[outputs[0]];
+        let strideW, strideH;
+        let filterW, filterH;
+        let activation;
+        if (inCount === 7) {
+          let paddingCode = operands[inputs[i++]].value[0];
+          let padding = PaddingCodeMap.get(paddingCode);
+          strideW = operands[inputs[i++]].value[0];
+          strideH = operands[inputs[i++]].value[0];
+          filterW = operands[inputs[i++]].value[0];
+          filterH = operands[inputs[i++]].value[0];
+          activation = FuseFunctionMap.get(operands[inputs[i++]].value[0]);
+          if (op === OperationCode.AVERAGE_POOL_2D) {
+            output.assign(activation(
+              input.avgPool([filterH, filterW], [strideH, strideW], padding)));
+          } else {
+            output.assign(activation(
+              input.maxPool([filterH, filterW], [strideH, strideW], padding)));
+          }
+        } else {
+          let paddingLeft = operands[inputs[i++]].value[0];
+          let paddingRight = operands[inputs[i++]].value[0];
+          let paddingTop = operands[inputs[i++]].value[0];
+          let paddingBottom = operands[inputs[i++]].value[0];
+          strideW = operands[inputs[i++]].value[0];
+          strideH = operands[inputs[i++]].value[0];
+          filterW = operands[inputs[i++]].value[0];
+          filterH = operands[inputs[i++]].value[0];
+          activation = FuseFunctionMap.get(operands[inputs[i++]].value[0]);
+          if (op === OperationCode.AVERAGE_POOL_2D) {
+            output.assign(activation(
+              input.pad([[0, 0], [paddingTop, paddingBottom], [paddingLeft, paddingRight], [0, 0]])
+                   .avgPool([filterH, filterW], [strideH, strideW], 'valid')));
+          } else {
+            output.assign(activation(
+              input.pad([[0, 0], [paddingTop, paddingBottom], [paddingLeft, paddingRight], [0, 0]])
+                   .maxPool([filterH, filterW], [strideH, strideW], 'valid')));
+          }
+        }
+      } break;
+      case OperationCode.SOFTMAX: {
+        let input = operands[inputs[0]];
+        let beta = operands[inputs[1]].value[0];
+        let output = operands[outputs[0]];
+        output.assign(input.mul(tf.scalar(beta)).softmax());
+      } break;
+      case OperationCode.RESHAPE: {
+        let input = operands[inputs[0]];
+        let targetShape = operands[inputs[1]];
+        let output = operands[outputs[0]];
+        output.assign(input.reshape(targetShape.dataSync()));
+      } break;
+      case OperationCode.CONCATENATION: {
+        if (outputs.length < 1 || inputs.length < 2) {
```

Comment:
simplify the condition?

Reply:
since tfjs will give detailed reasons for errors, can we remove this check?